### ISSUE TITLE
Tool development

### ISSUE THREAD

**Issue by yapple** [AppVet/appvet#3]
I want to development a tool web service for AppVet , can you give me some suggestion ? which tool ?

**Comment by aglynn** [comment]
One idea I would love to get help with is to automatically install the app on an emulator / simulator, click around the tool, and then return the network traffic information. That would be very helpful for users. Let me know if you are starting to work on this because maybe I can help you.

To develop other ideas, I would imagine that if you look through forensics or penetration testing manual tools, you might be able to find an innovative way to automate some of those. You might want to get comfortable with an emulator or simulator if you aren't already. I think that if you can get the app to be deployed onto your emulator or simulator you have a lot of ways to grab information that may be useful about the app. If you start developing some ideas, then I definitely help you flesh it out.

**Comment by yapple** [comment]
Yes , I'd like to do this . And I need you help .How would you help me ?

**Comment by aglynn** [comment]
If you create a public GitHub repository, then we can collaborate on it. I think the steps would be 1) get an emulator working on a server, 2) write the commands to install the apk on the emulator automatically, 3) leverage something to click around the application (I think ideally it would be an Appium script, or else a tool that randomly clicks around and fills out fields), 4) search the network logs for problematic strings (such as "id" or "password" or "http" (no "s")) and 5) return the logs back with a risk rating based on whether the problematic strings were found.

**Comment by yapple** [comment]
Do you know what kind of server we need, and what kind of simulator ?

**Comment by aglynn** [comment]
You'd probably want a Linux server, with an Android emulator since it is easier to get started with than the iOS simulator for remote automation. You'd want to download the android SDK and add the tools directory (where the "emulator" file is located) in order to get your "emulator" command working. Then you'd need to look up the command for installing the apk on the emulator. I don't know that command off the top of my head, but let me know whether you find that out.

**Comment by yapple** [comment]
I have now completed the first step and the second step, but the Appium framework is unfamiliar, the installation has not been successful, but I think the installation is not a problem. What I want to know is whether Appium provides a set of API that can be returned to the web log in the console. If it is, I don't need to think of ways to solve the problem. Do you have some help on Appium now?

**Comment by aglynn** [comment]
I am not the best at Appium unfortunately, but when you run your tests you should get the results from running those tests printed to the console. It should show whether the tests passed or failed, like with Selenium, I believe. What exactly would you like Appium to be printing out? The way I was thinking about it, we would just run an Appium test against an app and grab the network traffic from the emulator output, not really dealing much with the Appium results themselves, only to verify that they passed.

**Comment by yapple** [comment]
I learn a appium example, there is a question, whether there is a universal test script, or how to write a generic test script?

**Comment by aglynn** [comment]
Unfortunately, I don't think there is a generic script to use for that. It
depends on the app. If we used a basic flashlight app, that would probably
be easy to use. There are many flashlight apps out there that require
access to the internet because they are malicious. If on your or my local
computer we developed the Appium script, that shouldn't be too hard I
think. Let me know if you'd like me to do that. It is something I am
interested in doing more of. It might take me a few days to get that
developed though.

**Comment by yapple** [comment]
So we first developed a script for the flashlight application? How to get used to test the malicious program?

**Comment by aglynn** [comment]
You could maybe grab a flashlight app, such as by going to
https://apps.evozi.com/apk-downloader/?id=com.surpax.ledflashlight.panel,
clicking on the Generate Download Link, and click on "Click here to
download ... now". It'll download to your computer. Then you can do some of
the Appium steps to create the script for that, such as by
http://www.guru99.com/introduction-to-appium.html (look for the apk part)
or some other tutorial that might be better.


----

Anthony Glynn

**Comment by yapple** [comment]
I have now successfully run a appium test example, how are you going?

**Comment by aglynn** [comment]
Would you like me to do anything? I figured that when you get blocked and
when we try to integrate and make it more enterprise and reusable I'll get
my hands dirty with it.

**Comment by yapple** [comment]
I'm not sure what kind of test program we need?

**Comment by aglynn** [comment]
On the Appium documentation site it talks about how you can use various
languages to write the Appium test. "You can write tests with your favorite
dev tools using any WebDriver
<https://code.google.com/p/selenium/wiki/JsonWireProtocol>-compatible
language such as Java, Objective-C
<https://github.com/appium/selenium-objective-c>, JavaScript with Node.js
(inpromise <https://github.com/admc/wd>, callback
<https://github.com/admc/wd> or generator
<https://github.com/jlipps/yiewd> flavors),
PHP, Python, Ruby <https://github.com/appium/ruby_lib>, C#, Clojure, or
Perl with the Selenium WebDriver API and language-specific client
libraries." Are you now sure what language you should use? If so, then
Python, Ruby, PHP, Java, or Javascript would be my preference. Then we
would use it in top of the Android Emulator.

If I am not understanding the question, can you please clarify?

**Comment by yapple** [comment]
I have succeeded in running a Appium example using JAVA.
This is my code.
public class ContactsTest {
    private AndroidDriver<WebElement> driver; 
    @Before
    public void setUp() throws Exception {
    	
        //设置apk的路径
        File classpathRoot = new File(System.getProperty("user.dir"));
        System.out.println(System.getProperty("user.dir"));
        File appDir = new File(classpathRoot, "apps");
        File app = new File(appDir, "ContactManager.apk");
        
        //设置自动化相关参数
        DesiredCapabilities capabilities = new DesiredCapabilities();
        capabilities.setCapability(CapabilityType.BROWSER_NAME, "");
        capabilities.setCapability("platformName", "Android");
        capabilities.setCapability("deviceName", "Android Emulator");
        
        //设置安卓系统版本
        capabilities.setCapability("platformVersion", "4.3");
        //设置apk路径
        capabilities.setCapability("app", app.getAbsolutePath()); 
        
        //设置app的主包名和主类名
        capabilities.setCapability("appPackage", "com.example.android.contactmanager");
        capabilities.setCapability("appActivity", ".ContactManager");
        
        //初始化
        driver = new AndroidDriver<WebElement>(new URL("http://127.0.0.1:4723/wd/hub"), capabilities);       
    }
 
    @Test
    public void addContact(){
        WebElement el = driver.findElement(By.name("Add Contact"));
        el.click();
        List<WebElement> textFieldsList = driver.findElementsByClassName("android.widget.EditText");
        textFieldsList.get(0).sendKeys("Some Name");
        textFieldsList.get(2).sendKeys("Some@example.com");
        driver.swipe(100, 500, 100, 100, 2);
        driver.findElementByName("Save").click();
    }    
    
    @After
    public void tearDown() throws Exception {
        driver.quit();
    }
}
But I don't know how to write a more generic code

**Comment by aglynn** [comment]
For now, if this code works for this, we do not need more generic code. The
way I envision it us that ultimately the new AppVet will tool we are
building will pull from a Git repos that have Appium tests and will pull
the right repo based on the app package name or some other identifier.

Does your Appium test work for this app? If so, is your AppVet tool able to
trigger that test to run for the app? Can you please commit your tool code
to GitHub for your Appvet tool so I can check it out and help with it?
